### ISSUE TITLE
18.06 failed reauthentification loops

### ISSUE THREAD

**Issue by juliantaylor** [gocd/gocd#4872]
##### Issue Type

- Feature proposal
- Regression

##### Summary

Gocd 18.06 reauthentifcation loops, allow disabling reauthentification or add a backoff on failure.

Our gocd server uses ldap authentification with the ldap using HOTP 2 factor authentication. The re-authentication gocd uses cannot work with this setup so we have increased the reauthentification interval to several hours which worked fine in 17.12.
Since upgrading to gocd 18.06 when this interval is exceeded and the user still has a browser window with gocd open the server starts authenticate about 20 times every 10 seconds effectively dos'ing our ldap authentication.
We will probably increase the interval to some effectively infinite time (or can you disable it completely?) but gocd should have a saner behavior for this situation, for example have a increasing backoff time when the authentificatino fails to not overload the backend.

**Comment by bdpiparva** [comment]
@juliantaylor -- Please can you provide more detail about the issue

1. Is this issue faced by all user's having GoCD window open in the browser?
2. This happens to a selected user for whom LDAP credentials are changed?
   - For this, there is an open issue https://github.com/gocd/gocd/issues/4853 which will be fixed in PR https://github.com/gocd/gocd/pull/4868

In case if you are facing the first issue please can you share plugin and go-server log.

**Comment by juliantaylor** [comment]
It appears to happen for all users but the credentials becoming invalid due to the OTP could counts as changed credentials? We are also seeing the redirect loop which has to be fixed by deleting the cookies.

the plugin-cd.go.authentication.ldap logs are full with these types of logs:

```
2018-06-21 12:42:27,955 INFO  [qtp1489069835-23] LdapPlugin:73 - [Authenticate] Authenticating User: aafafwgeo using auth_config: ed972eee-d2c0-4ab7-bf2b-3905e3559cb6
2018-06-21 12:42:28,207 ERROR [qtp1489069835-23] LdapPlugin:128 - [Authenticate] Failed to authenticate user aafafwgeo on ldaps://ldapserver
```

Anything in particular you want from the go-server logs?

**Comment by ketan** [comment]
> It appears to happen for all users but the credentials becoming invalid due to the OTP could counts as changed credentials

If you sending your LDAP password + OTP in the password field — then that would be considered as changed credentials.

**Comment by ketan** [comment]
Could you confirm if this is a new installation — and if its an upgrade of GoCD, what version of GoCD you were on earlier?

**Comment by bdpiparva** [comment]
> It appears to happen for all users but the credentials becoming invalid due to the OTP could counts as changed credentials? We are also seeing the redirect loop which has to be fixed by deleting the cookies.

Yes, [redirection loop](https://github.com/gocd/gocd/issues/4853) bug is introduced with GoCD release 18.6.0  and it will be fixed in next release. 

PR https://github.com/gocd/gocd/pull/4868 will redirect the user to login page when credentials changes.

Here are some workaround for the issue https://github.com/gocd/gocd/issues/3779#issuecomment-324863860

**Comment by juliantaylor** [comment]
it is not a new installation, it has been updated several times from atleast 17.08 to 17.12 to now 18.06

We have disabled the reauth and set go.server.session.timeout.second and  go.sessioncookie.maxage.seconds to 10h, maybe that helps.

**Comment by ketan** [comment]
@juhijariwala — could you please confirm if the build pointed above fixes the problem at your end?

**Comment by juliantaylor** [comment]
the settings I mentioned above fixed the problem for 18.6
I have installed 18.7 and set the timeouts to a couple minutes and could not reproduce the reauth loop so I assume it is fixed.

**Comment by blzysh** [comment]
I've tried this build and do not get the redirects but get frequent (sessions) timeouts.

@juliantaylor - can you explain where you're setting these options?

**Comment by juliantaylor** [comment]
@blzysh these are options set during gocd startup, see https://github.com/gocd/gocd/issues/3779#issuecomment-324863860
